Support *args in Pipe.forward API.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57325 Add `NoChunk` wrapper for pipeline args.
* #57226 Support non-Tensor args in the Pipe API
* **#55441 Support *args in Pipe.forward API.**

This is the first step towards supporting the proposal outlined in
https://github.com/pytorch/pytorch/issues/53952.

In this PR I've ensured Pipe.forward() accepts a *inputs argument instead of
just a single input as previously. This lays the groundwork for supporting
non-Tensors and generic arguments to the Pipe API. In this PR we still only
support Tensors and non-Tensor support will come in future PRs.

For backward compatibility I've ensured a single Tuple[Tensor] input still
works as expected previously.

Differential Revision: [D27613887](https://our.internmc.facebook.com/intern/diff/D27613887/)

pritamdamania committed:
Support *args in Pipe.forward API.

This is the first step towards supporting the proposal outlined in
https://github.com/pytorch/pytorch/issues/53952.

In this PR I've ensured Pipe.forward() accepts a *inputs argument instead of
just a single input as previously. This lays the groundwork for supporting
non-Tensors and generic arguments to the Pipe API. In this PR we still only
support Tensors and non-Tensor support will come in future PRs.

For backward compatibility I've ensured a single Tuple[Tensor] input still
works as expected previously.

Differential Revision: [D27613887](https://our.internmc.facebook.com/intern/diff/D27613887/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support *args in Pipe.forward API."


This is the first step towards supporting the proposal outlined in
https://github.com/pytorch/pytorch/issues/53952.

In this PR I've ensured Pipe.forward() accepts a *inputs argument instead of
just a single input as previously. This lays the groundwork for supporting
non-Tensors and generic arguments to the Pipe API. In this PR we still only
support Tensors and non-Tensor support will come in future PRs.

For backward compatibility I've ensured a single Tuple[Tensor] input still
works as expected previously.

Differential Revision: [D27613887](https://our.internmc.facebook.com/intern/diff/D27613887/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support *args in Pipe.forward API."


This is the first step towards supporting the proposal outlined in
https://github.com/pytorch/pytorch/issues/53952.

In this PR I've ensured Pipe.forward() accepts a *inputs argument instead of
just a single input as previously. This lays the groundwork for supporting
non-Tensors and generic arguments to the Pipe API. In this PR we still only
support Tensors and non-Tensor support will come in future PRs.

For backward compatibility I've ensured a single Tuple[Tensor] input still
works as expected previously.

Differential Revision: [D27613887](https://our.internmc.facebook.com/intern/diff/D27613887/)

[ghstack-poisoned]

mrzzd commented on torch/distributed/pipeline/sync/microbatch.py:
```diff
@@ -132,39 +141,44 @@ def _setitem_by_slice(self, index: slice, value: Tensors) -> None:
-                raise TypeError(f"expected Tensor, but got {input.__class__.__name__}")
-        return
+    for input in inputs:
+        if isinstance(input, Sequence):
```
We allow Sequence only if it `len(inputs)==1` , right?

mrzzd commented on torch/distributed/pipeline/sync/pipe.py:
```diff
@@ -248,8 +240,9 @@ def __init__(
 
-        self.partitions, self.devices = _split_module(module)
-        _verify_splitting(module, self.partitions, self.devices)
+        self.partitions = module
```
could you elaborate how this change helps what you wanted to achieve in this PR? with _split_module, we had the flexibility of changing partitioning algorithm by modifying _split_module implementation.

pritamdamania87 replied:
`split_module` was mostly redundant since we don't really split the module and in this API the module is already split by the user. Since this is an internal implementation detail, we can always change it later if needed.

pritamdamania committed:
Update on "Support *args in Pipe.forward API."


This is the first step towards supporting the proposal outlined in
https://github.com/pytorch/pytorch/issues/53952.

In this PR I've ensured Pipe.forward() accepts a *inputs argument instead of
just a single input as previously. This lays the groundwork for supporting
non-Tensors and generic arguments to the Pipe API. In this PR we still only
support Tensors and non-Tensor support will come in future PRs.

For backward compatibility I've ensured a single Tuple[Tensor] input still
works as expected previously.

Differential Revision: [D27613887](https://our.internmc.facebook.com/intern/diff/D27613887/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support *args in Pipe.forward API."


This is the first step towards supporting the proposal outlined in
https://github.com/pytorch/pytorch/issues/53952.

In this PR I've ensured Pipe.forward() accepts a *inputs argument instead of
just a single input as previously. This lays the groundwork for supporting
non-Tensors and generic arguments to the Pipe API. In this PR we still only
support Tensors and non-Tensor support will come in future PRs.

For backward compatibility I've ensured a single Tuple[Tensor] input still
works as expected previously.

Differential Revision: [D27613887](https://our.internmc.facebook.com/intern/diff/D27613887/)

[ghstack-poisoned]

mrzzd commented on torch/distributed/pipeline/sync/pipeline.py:
```diff
@@ -82,7 +82,7 @@ class Pipeline:
     def __init__(
         self,
-        partitions: List[nn.Sequential],
+        partitions: nn.Sequential,
```
partitions seems to be list not a module

mrzzd commented on torch/distributed/pipeline/sync/pipe.py:
```diff
@@ -308,7 +329,10 @@ def __getitem__(self, index: int) -> nn.Module:
         """Iterates over children of the underlying sequential module."""
         for partition in self.partitions:
-            yield from partition
+            if isinstance(partition, nn.Sequential):
```
is this change needed?

mrzzd commented on torch/distributed/pipeline/sync/pipe.py:
```diff
@@ -291,12 +307,17 @@ def __getitem__(self, index: int) -> nn.Module:
-                pass
 
-            shift = len(partition)
+            if isinstance(partition, nn.Sequential):
```
are these changes needed?

mrzzd commented on torch/distributed/pipeline/sync/pipe.py:
```diff
@@ -282,7 +298,7 @@ def __init__(
     def __len__(self) -> int:
         """Counts the length of the underlying sequential module."""
-        return sum(len(p) for p in self.partitions)
+        return sum(len(p) if isinstance(p, nn.Sequential) else 1 for p in self.partitions)
```
revert this?

mrzzd commented on torch/distributed/pipeline/sync/pipe.py:
```diff
@@ -68,7 +68,7 @@ def _verify_module(module: nn.Sequential) -> None:
 
 def _verify_splitting(
-    module: nn.Sequential, partitions: List[nn.Sequential], devices: List[torch.device]
+    module: nn.Sequential, partitions: nn.Sequential, devices: List[torch.device]
```
revert this?

mrzzd commented on torch/distributed/pipeline/sync/microbatch.py:
```diff
@@ -40,9 +40,11 @@ class Batch:
         self.value = value
         self.atomic = torch.is_tensor(value)
+        # For backward compatibility.
+        self.is_single_sequence = is_single_sequence
```
how `is_single_sequence` is different from `atomic`?

mrzzd replied:
Sounds good. Maybe it would be better if you add a description of `is_single_sequence` to the docstring

pritamdamania committed:
Update on "Support *args in Pipe.forward API."


This is the first step towards supporting the proposal outlined in
https://github.com/pytorch/pytorch/issues/53952.

In this PR I've ensured Pipe.forward() accepts a *inputs argument instead of
just a single input as previously. This lays the groundwork for supporting
non-Tensors and generic arguments to the Pipe API. In this PR we still only
support Tensors and non-Tensor support will come in future PRs.

For backward compatibility I've ensured a single Tuple[Tensor] input still
works as expected previously.

Differential Revision: [D27613887](https://our.internmc.facebook.com/intern/diff/D27613887/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support *args in Pipe.forward API."


This is the first step towards supporting the proposal outlined in
https://github.com/pytorch/pytorch/issues/53952.

In this PR I've ensured Pipe.forward() accepts a *inputs argument instead of
just a single input as previously. This lays the groundwork for supporting
non-Tensors and generic arguments to the Pipe API. In this PR we still only
support Tensors and non-Tensor support will come in future PRs.

For backward compatibility I've ensured a single Tuple[Tensor] input still
works as expected previously.

Differential Revision: [D27613887](https://our.internmc.facebook.com/intern/diff/D27613887/)

[ghstack-poisoned]

mrzzd commented on test/distributed/pipeline/sync/test_microbatch.py:
```diff
@@ -44,7 +44,7 @@ def test_batch_non_atomic():
 def test_batch_call():
     a = Batch(torch.tensor(42))
-    b = Batch((torch.tensor(42), torch.tensor(21)))
+    b = Batch((torch.tensor(42), torch.tensor(21)), is_single_sequence=True)
```
it seems we do not have a similar test with is_single_sequence=False (that tests .call method) here.

mrzzd commented on torch/distributed/pipeline/sync/microbatch.py:
```diff
@@ -132,39 +141,44 @@ def _setitem_by_slice(self, index: slice, value: Tensors) -> None:
+        unwrapped_inputs = zip(*rotated)  # type: ignore[assignment]
 
-    return [Batch(x) for x in inputs]
+    return [Batch(x, is_single_sequence) for x in unwrapped_inputs]
```
Suppose the user creates `Pipe` for `nn.Sequence(m1: M1, m2: M2)` with following signature:
```py
class M1(nn.Module):
   def forward(self, input: Tensor) ->  Tuple[Tensor, Tensor]:
        ...
class M2(nn.Module):
   def forward(self, input: Tuple[Tensor, Tensor]) ->  Tensor:
        ...
```
This works fine with old implementation, but it seems to me with new implementation we set is_single_sequence=False, which is wrong.

pritamdamania87 replied:
@mrzzd I thought about this for a while and couldn't come up with a good solution. I did consider your proposal about adding `is_single_sequence` to the constructor. The tricky thing with that is that if by default that is False, we anyways end up breaking compatibility since the user has to manually set this flag to get the old behavior.

I felt it was probably better to just break compatibility since the feature is Beta and keep the behavior simple and consistent without special handling for tuple of tensors. We can call this out in our release notes as well for the next release. As a result, I removed most of the logic pertaining to is_single_sequence.

cc @mrshenli

pritamdamania committed:
Update on "Support *args in Pipe.forward API."


This is the first step towards supporting the proposal outlined in
https://github.com/pytorch/pytorch/issues/53952.

In this PR I've ensured Pipe.forward() accepts a *inputs argument instead of
just a single input as previously. This lays the groundwork for supporting
non-Tensors and generic arguments to the Pipe API. In this PR we still only
support Tensors and non-Tensor support will come in future PRs.

For backward compatibility I've ensured a single Tuple[Tensor] input still
works as expected previously.

Differential Revision: [D27613887](https://our.internmc.facebook.com/intern/diff/D27613887/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support *args in Pipe.forward API."


This is the first step towards supporting the proposal outlined in
https://github.com/pytorch/pytorch/issues/53952.

In this PR I've ensured Pipe.forward() accepts a *inputs argument instead of
just a single input as previously. This lays the groundwork for supporting
non-Tensors and generic arguments to the Pipe API. In this PR we still only
support Tensors and non-Tensor support will come in future PRs.

For backward compatibility I've ensured a single Tuple[Tensor] input still
works as expected previously.

Differential Revision: [D27613887](https://our.internmc.facebook.com/intern/diff/D27613887/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support *args in Pipe.forward API."


This is the first step towards supporting the proposal outlined in
https://github.com/pytorch/pytorch/issues/53952.

In this PR I've ensured Pipe.forward() accepts a *inputs argument instead of
just a single input as previously. This lays the groundwork for supporting
non-Tensors and generic arguments to the Pipe API. In this PR we still only
support Tensors and non-Tensor support will come in future PRs.

For backward compatibility I've ensured a single Tuple[Tensor] input still
works as expected previously.

Differential Revision: [D27613887](https://our.internmc.facebook.com/intern/diff/D27613887/)

[ghstack-poisoned]

SciPioneer commented on torch/distributed/pipeline/sync/microbatch.py:
```diff
@@ -132,39 +135,32 @@ def _setitem_by_slice(self, index: slice, value: Tensors) -> None:
 
 
-def check(input: TensorOrTensors) -> None:
+def check(*inputs) -> None:
```
nit: no type hint?

pritamdamania87 replied:
The PRs on top allow for generic inputs, as a result I didn't put a type hint. This is similar to *args.

SciPioneer commented on torch/distributed/pipeline/sync/microbatch.py:
```diff
@@ -132,39 +135,32 @@ def _setitem_by_slice(self, index: slice, value: Tensors) -> None:
 
-
-def scatter(input: TensorOrTensors, chunks: int) -> List[Batch]:
+def scatter(*inputs, chunks: int) -> List[Batch]:
```
nit: no type hint?

pritamdamania87 replied:
Same as above.

wayi1 reviewed (approved):
LGTM, only one question about the type hint.

mrzzd commented on test/distributed/pipeline/sync/test_pipe.py:
```diff
@@ -650,3 +636,19 @@ def forward(self, x):
+        def forward(self, a, b):
+            return a + b
+
+    model = Pipe(nn.Sequential(Module1().cuda(0), Module2().cuda(0)), chunks=2, checkpoint=checkpoint)
```
should change to PipeSequential

pritamdamania87 replied:
PipeSequential should only be used internally inside the Pipe API right? I would not expect users to use PipeSequential.

mrzzd reviewed (approved):
Thanks Pritam.
